[1주차 기본 과제] Todo List 만들기

myeongheonhong opened the pull request:
<!-- PR의 제목은 "[ n주차 기본/심화/생각 과제 ] 과제명 " 으로 작성해주시면 되겠습니다 -->
<!-- PR은 리뷰어를 위한 글입니다, 보다 더 상세하게 적음으로써 소통을 활성화해봐요! -->

## ✨ 구현 기능 명세

- 기본 과제
1. header, nav, main (left side, right side) / footer 를 나누어 구현합니다
2. nav 에서는 3개의 버튼이 있어요 (가운데에 오도록 배치해주세요)
3. main 에는 투두 리스트를 위치시켜요
* 두 개의 section 으로 나누어 left, right 를 양 쪽에 배치해요
* 오늘 할 일, 내일 할 일은 헤딩 태그를 / 할 일 리스트는
  - 태그를 이용해요
  - 내일 할 일에는 2개의 리스트가 존재합니다 ! 각 리스트의 구분선과, 삭제 버튼이 있어요
  - section 의 아래에는 input 태그와 추가 버튼이 존재해요

<br />

## 🎁 PR Point

태그들을 정렬하는 부분과 일정한 웹뷰를 보여주고자 스타일에 신경써서 구현했습니다.

<br />

## 😭 소요 시간, 어려웠던 점

- `2h`
- main 부분을 화면에 꽉차게 보여주고자 하는 부분에서 많은 시간을 쓴거 같습니다 ㅠㅠ

<br />

## 😎 구현 결과물
<img width="1552" alt="스크린샷 2022-10-04 오후 4 56 12" src="https://user-images.githubusercontent.com/70846061/193765598-75fa2197-1de0-4ccd-a1cd-3666e1137159.png">

<!-- 스크린샷, gif, 배포링크 등 자유롭게 작성하되 애니메이션이나 동적UI 확인이 필요한 경우에는 스크린샷은 인정하지 않습니다. -->

myeongheonhong committed:
init : 초기파일 생성

myeongheonhong committed:
init : html 레이아웃 작업 및 css파일 생성.

myeongheonhong committed:
css : header, nav, main, section, footer 스타일 적용.

leeseooo commented on Week01/index.html:
```diff
@@ -4,51 +4,77 @@
+                <button class="btn">내일만 보기</button>
+                <button class="btn">함께 보기</button>
             </div>
         </nav>
```
굳이 div로 한 번 더 감싸지 않고 `nav` 태그에 class 줘서 버튼 중앙 정렬해도 좋을 것 같아요~

leeseooo commented on Week01/index.html:
```diff
@@ -4,51 +4,77 @@
 
         <main>
             <!-- LEFT SECTION -->
             <section id="left-section">
```
웹표준에 따르면... `section` 태그 안에는 h1과 같은 헤딩 태그가 필요하다고 하네요

참고 - https://developer.mozilla.org/ko/docs/Web/HTML/Element/section
```
각각의 <section>을 식별할 수단이 필요합니다. 
주로 제목<h1>-<h6>요소를 <section>의 자식으로 포함하는 방법을 사용합니다.
```

leeseooo commented on Week01/index.html:
```diff
@@ -4,51 +4,77 @@
+                        <li>2. 모각공하기.</li>
+                        <span class="material-icons"> cancel </span>
+                    </div>
                 </ul>
```
모각공 때 말했던 ul 안에 div 태그가 이 말이었군요~ㅎㅎ
이런 경우에는 div와 span은 빼고 li태그에 가상요소(::after, ::before)를 이용해서 아이콘을 넣어도 될 것 같아요.
저렇게 하면 모든 li태그에 div로 감싸고 span으로 아이콘을 넣어줘야 해서 코드가 길어질거에요.

```css
li::after{
  content:'';
  width:2.4rem;
  height:2.4rem;
  background-size: 2.4rem;
  background-image: url('./close.svg');
}
```
원하는 아이콘을 link로 import하지 말고, 직접 다운받아서 가상요소의 background-image: url( ) 안에 넣어주면 좋을 것 같아요.

leeseooo commented on Week01/style.css:
```diff
@@ -0,0 +1,103 @@
+    display: flex;
+    justify-content: space-between;
+    align-items: center;
+    margin: 10px 10px;
```
요건 margin: 10px로 쓸 수 있겠네요

leeseooo reviewed (commented):
굿👍

Happhee commented on Week01/index.html:
```diff
@@ -4,7 +4,58 @@
+                <button>오늘만 보기</button>
+                <button>내일만 보기</button>
+                <button>함께 보기</button>
+            </div>
```
여기 안에도 div 태그 보다는 section태그가 적합하지 아느까? button들을 모아둔 구역이니까!

myeongheonhong replied:
@Happhee 혹시 그런정보들을 어디서 얻을 수 있을까....?!

Happhee commented on Week01/index.html:
```diff
@@ -4,7 +4,58 @@
+                <div class="add">
+                    <input type="text" />
+                    <button>+</button>
+                </div>
```
여기서도 살짝 시멘틱구조로 바꾸면 위에 li들이 input과 button에 대한 사용자의 입력 및 전송으로 목록들이 추가데자나! 그래서 [form태그](https://www.nextree.co.kr/p8428/)가 적합할듯해!! 

```suggestion
 <form  class="add-form" action="http..." accept-charset="utf-8" name="todos_info" method="post" >
       <input type="text" />
        <button>+</button>
</form>
``` 
폼속성이 뭐가 있는지도 공부해봤으면 좋겟당 ㅎㅎ

Happhee commented on Week01/index.html:
```diff
@@ -4,51 +4,77 @@
-                <button>오늘만 보기</button>
-                <button>내일만 보기</button>
-                <button>함께 보기</button>
+                <button class="btn">오늘만 보기</button>
+                <button class="btn">내일만 보기</button>
+                <button class="btn">함께 보기</button>
```
클래스명이랑 태그이름의 뜻이 비슷한데 혹시 그렇게 한 이유가 있을까?! 버튼에 type도 공부해보구~.~

Happhee commented on Week01/index.html:
```diff
@@ -4,51 +4,77 @@
-                    <li>1</li>
-                    <li>2</li>
+                    <h1 class="title">내일 할일</h1>
+                    <div class="item">
+                        <li>1. 알바가기</li>
+                        <span class="material-icons"> cancel </span>
+                    </div>
```
아니면 어차피 ul 안에 li가 나열이고 li태그가 블록요소니까 하나로 묶을수도 있을것같아!
클래스명도 item이라서 !!
```suggestion
 <li class="right-section__item">1. 알바가기  <span class="material-icons"> cancel </span>
 </li>
                      
```

leeseooo replied:
이 방법도 좋네요~

Happhee commented on Week01/style.css:
```diff
@@ -0,0 +1,103 @@
+    border: none;
+}
+
+.btn {
+    border-radius: 10px;
+    background-color: #dea01e;
+    width: 100px;
+    height: 35px;
+    cursor: pointer;
+}
+
+/* HEADER */
+header {
+    height: 100px;
+    background-color: #3e87cb;
+    display: flex;
+    justify-content: center;
+    align-items: center;
+}
+
```
[css 속성순서](http://milooy.github.io/TIL/CSS/css-property-order.html#mozilla)를 지켜서 쓰면 나중에 협업할때 편할거야~

myeongheonhong replied:
@Happhee 이거 진짜 너무 꿀팁이다.....ㅠㅠㅠ정말 고마웡ㅎㅎㅎ

Happhee commented on Week01/style.css:
```diff
@@ -0,0 +1,103 @@
+    margin: 10px 10px;
+}
+
+main section .add {
+    display: flex;
+    justify-content: center;
+    margin-bottom: 10px;
+}
```
클래스명 하나만 써도 되는데 여러개를 나열한 이유가 있을까?

Happhee commented on Week01/style.css:
```diff
@@ -0,0 +1,103 @@
+    border: solid;
+}
+
+main section .add button {
+    background-color: transparent;
+    cursor: pointer;
+}
```
여기서는 main section .add 밑의 button이니까 나중에 혼돈생기지 않게 선택자를 활용해봤으면 좋게써 ㅎㅅㅎ

Happhee reviewed (commented):
레이아웃 맞추는게 힘들긴하지...ㅠㅠ 프론트엔드에게 가장 디테일한부분인것가타 ㅎㅅㅎ 
class명이랑 우리 웹접근성에대해 이번에 생각해보는게 과제였으니까 시멘틱구조로 짜는걸 좀더 노력해봤으면 좋겟어!!  그게 집중하지 않으면 생각보다 많이 놓치게 되더라구 ㅠㅠ

고생해따~.~

NaveOWO commented:
구경왔는데 서영이가 아주 꼼꼼보스다🌟🌟 나도 화면 레이아웃 내맘대로 안될때 너무 힘든데, 나는 그럴 때 px말고 %를 사용해서 화면에 보여주고싶은 콘텐츠의 크기를 정하기도 해!! 멋지당!!!!!수고햇오!

myeongheonhong committed:
[Fix] : 코드리뷰 피드백을 반영한 코드수정

myeongheonhong committed:
[css] : 중간 button 간격조정(피드백 반영!!)

myeongheonhong committed:
feat : 함수정의.

myeongheonhong committed:
feat :
1. nav-button별로 오늘만 보기, 내일만 보기, 함께보기 함수 구현,
2. 오늘할일과 내일할일의 리스트 추가기능 함수 구현.
3. 리스트 삭제하는 기능 구현.(다시 생성되는 목록에는 적용이 안되는 버그가 있음...ㅠ)

myeongheonhong committed:
feat : 동적으로 생성된 할 일 리스트에도 삭제하는 이벤트를 추가.

leeseooo commented on Week01/main.js:
```diff
@@ -0,0 +1,76 @@
+    right_section.className = '';
+}
+
+//반복문을 통해 navigation button에 이벤트 할당.
+for (let i = 0; i < btn_group.length; i++) {
+    btn_group[i].addEventListener('click', () => {
+        buttonByAttribute(btn_group[i]);
+    });
+}
```
반복문으로 for문을 사용해도 되지만, `forEach`같은 배열 순회 메서드를 사용해보는 건 어떨까요?
저도 요즘 선언형 프로그래밍을 연습하고 있어요.

[참고 - 명령형 프로그래밍 vs 선언형 프로그래밍](https://sia-world.tistory.com/entry/%EB%AA%85%EB%A0%B9%ED%98%95-%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D-vs-%EC%84%A0%EC%96%B8%ED%98%95-%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D?category=1083237)

leeseooo commented on Week01/main.js:
```diff
@@ -0,0 +1,76 @@
+    }
+}
+
+//한쪽만 보이게 하는 함수
+function oneSideView(attr) {
+    if (attr === 'today') {
+        right_section.className = 'hidden';
+        left_section.className = '';
+    } else {
+        right_section.className = '';
+        left_section.className = 'hidden';
+    }
+}
```
깔끔하네요 ㅎㅎ 👍👍👍

leeseooo commented on Week01/main.js:
```diff
@@ -0,0 +1,76 @@
+//할 일 추가 함수.
+function addList(time) {
+    const li = document.createElement('li');
+    const textNode = document.createTextNode(todayInput.value);
+    li.appendChild(textNode);
```
`createTextNode`라는 게 있군요.. 👍👍👍
요소의 텍스트를 변경하기 위해 `innerText` 속성을 사용할 수도 있어요.

leeseooo reviewed (commented):
가독성이 좋고 함수 분리도 깔끔하게 해주신 것 같아요 ㅎㅎ
고생하셨어요!
